htmresearch/algorithms/column_pooler.py

```diff
@@ -83,319 +75,389 @@ def __init__(self,
-                      columnDimensions=columnDimensions,
-                      basalInputDimensions=(lateralInputWidth,),
-                      apicalInputDimensions=(),
-                      cellsPerColumn=1,
-                      activationThreshold=activationThresholdDistal,
-                      initialPermanence=initialPermanence,
-                      connectedPermanence=connectedPermanence,
-                      minThreshold=minThresholdDistal,
-                      maxNewSynapseCount=maxNewDistalSynapseCount,
-                      permanenceIncrement=permanenceIncrement,
-                      permanenceDecrement=permanenceDecrement,
-                      predictedSegmentDecrement=predictedSegmentDecrement,
-                      formInternalBasalConnections=False,
-                      learnOnOneCell=False,
-                      maxSegmentsPerCell=maxSegmentsPerCell,
-                      maxSynapsesPerSegment=maxSynapsesPerDistalSegment,
-                      seed=seed,
-    )
-
-    # These sparse matrices will hold the synapses for each proximal segment.
-    #
-    # proximalPermanences - SparseMatrix with permanence values
-    # proximalConnections - SparseBinaryMatrix of connected synapses
-
-    self.proximalPermanences = SparseMatrix(self.numberOfColumns(),
-                                               inputWidth)
-    self.proximalConnections = SparseBinaryMatrix(inputWidth)
-    self.proximalConnections.resize(self.numberOfColumns(), inputWidth)
-
-
-  def depolarizeCells(self, activeExternalCells, learn=True):
-    """
-    Parameters:
-    ----------------------------
-    @param  activeExternalCells  (set)
-            Indices of active cells that will form connections to distal
-            segments.
+    @param  connectedPermanenceProximal (float)
+            Permanence required for a proximal synapse to be connected
 
-    @param  learn (bool)
-            If true, distal segment activations will be recorded. This
-            information is used during segment cleanup.
+    @param  synPermDistalInc (float)
+            Permanence increment for distal synapses
 
-    """
-    self.tm.depolarizeCells(activeCellsExternalBasal=activeExternalCells,
-                            learn=learn)
+    @param  synPermDistalDec (float)
+            Permanence decrement for distal synapses
 
+    @param  sampleSizeDistal (int)
+            Number of distal synapses a cell should grow to each lateral
+            pattern, or -1 to connect to every active bit
 
-  def activateCells(self,
-                    feedforwardInput=(),
-                    reinforceCandidatesExternal=(),
-                    growthCandidatesExternal=(),
-                    learn=True):
-    """
+    @param  initialDistalPermanence (float)
+            Initial permanence value for distal synapses
 
-    @param  feedforwardInput (set)
-            Indices of active input bits
+    @param  minThresholdDistal (int)
```

Comment:
In TemporalMemory terminology, these are both "activation" thresholds. They're not TM "minThresholds" (which should probably be called "learning threshold")

In this implementation, a cell devotes exactly 0 or 1 segments to each L2. When a cell appears in multiple SDRs, or if it learns multiple SDRs in a single lateral input, it will use the same segment for all of them. So having separate "learning" and "activation" thresholds isn't meaningful. This just has the "activation" threshold. I copied the proximal parameter names, but we could choose to change them both to "activationThreshold".

If we want one segment per pattern, we'd have to use the Connections. Using this one-segment approach, I don't think we'll see any scaling issues. We're putting much less stress on the distal segments than we are on the proximal.